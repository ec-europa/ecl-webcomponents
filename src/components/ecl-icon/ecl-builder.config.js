const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, '../../../node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  copy: [
    {
      from: [path.resolve(nodeModules, '@ecl/resources-social-media-icons/dist/sprites/icons-social-media.svg')],
      to: path.resolve(outputFolder, 'images/ec'),
      options: { up: true },
    },
    {
      from: [path.resolve(nodeModules, '@ecl/resources-social-media-icons/dist/sprites/icons-social-media.svg')],
      to: path.resolve(outputFolder, 'images/eu'),
      options: { up: true },
    },
    {
      from: [path.resolve(nodeModules, '@ecl/resources-ec-icons/dist/sprites/icons.svg')],
      to: path.resolve(outputFolder, 'images/ec'),
      options: { up: true },
    },
    {
      from: [path.resolve(nodeModules, '@ecl/resources-eu-icons/dist/sprites/icon.svg')],
      to: path.resolve(outputFolder, 'images/eu'),
      options: { up: true },
    },
  ],
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-icon-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-icon-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-icon-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-icon-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
  ],
};
