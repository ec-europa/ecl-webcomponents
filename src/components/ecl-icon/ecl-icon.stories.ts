import iconPathEc from '@ecl/resources-ec-icons/dist/sprites/icons.svg';
import iconPathEu from '@ecl/resources-eu-icons/dist/sprites/icons.svg';
import iconsAllEc from '@ecl/resources-ec-icons/dist/lists/all.json';

export default {
  title: 'Components/icon',
};

const Template = (args) => `<ecl-icon 
                              icon="${args.icon}"
                              color="${args.color}"
                              size="${args.size}"
                              transform="${args.transform}"
                              path="${args.theme === 'ec' ? iconPathEc : iconPathEu}"
                            ></ecl-icon>`;

export const Icon = Template.bind({});
Icon.storyName = 'default';
Icon.args = {
  theme: 'ec',
  icon: 'close',
  transform: '',
  color: 'default',
  size: '2xl',
};
Icon.argTypes = {
  theme: {
    type: { name: 'select'},
    options: ['ec', 'eu'],
  },
  icon: {
    type: { name: 'select' },
    options: iconsAllEc,
    description: "Add an icon",
  },
  size: {
    type: { name: 'select' },
    options: ['2xs','xs', 's', 'm', 'l', 'xl', '2xl'],
    description: "icon size",
  },
  color: {
    type: { name: 'select' },
    options: ['default', 'primary', 'inverted'],
    description: "icon color",
  },
  transform: {
    type: { name: 'select' },
    description: 'Button icon transform',
    options: [
      'rotate-90',
      'rotate-180',
      'rotate-270',
      'flip-horizontal',
      'flip-vertical',
    ],
  },
};