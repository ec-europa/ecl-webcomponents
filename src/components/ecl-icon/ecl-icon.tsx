import { Component, Prop, h } from '@stencil/core';
import getAssetPath from "../../utils/assetPath";

@Component({
  tag: 'ecl-icon',
  styleUrls: {
    ec: './build/styles/ecl-icon-ec.css',
    eu: './build/styles/ecl-icon-eu.css',
  },
  shadow: false,
  scoped: true,
  assetsDirs: ['build'],
})

export class EclIcon {
  @Prop() styleClass: string = '';
  @Prop() icon: string = '';
  @Prop() size: string = 'xs';
  @Prop() color: string = '';
  @Prop() titleTag: string = '';
  @Prop({ mutable: true }) path: string;
  @Prop() transform: string = '';
  @Prop({ mutable: true }) theme: string;
  @Prop() sprite: string = '';

  componentWillLoad() {
    this.theme = document.documentElement.getAttribute('data-ecl-theme') ?? (this.theme || 'ec');
    
    if (!this.sprite) {
      this.path = getAssetPath(`./build/images/${this.theme}/icons.svg`);
    } else {
      this.path = getAssetPath(`./build/images/${this.theme}/${this.sprite}.svg`);
    }
  }

  getClass(): string {
    const styleClasses = [
      `ecl-icon`,
      `ecl-icon--${this.size}`,
      this.styleClass,
    ];
    if (this.color) {
      styleClasses.push(
        `ecl-icon--${this.color}`
      );
    }
    if (this.transform) {
      styleClasses.push(
        `ecl-icon--${this.transform}`
      );
    }

    return styleClasses.join(' ');
  }

  render() {
    return (
      <svg class={this.getClass()}>
      { this.titleTag ? <title>{this.titleTag}</title> : '' } 
        <use xlinkHref={`${this.path}#${this.icon}`}></use>
      </svg>
    )
  }
}