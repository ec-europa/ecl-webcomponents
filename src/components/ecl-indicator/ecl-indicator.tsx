import { Component, Prop, h, Element } from '@stencil/core';
import getAssetPath from '../../utils/assetPath';
declare const ECL: any;

@Component({
  tag: 'ecl-indicator',
  styleUrls: {
    ec: './build/styles/ecl-indicator-ec.css',
    eu: './build/styles/ecl-indicator-eu.css',
  },
  shadow: false,
  scoped: true,
  assetsDirs: ['build'],
})
export class EclIndicator {
  @Element() el: HTMLElement;
  @Prop() value: string = '';
  @Prop() extraClasses: string = '';
  @Prop() eclScript: boolean = false;
  @Prop() styleClass: string = '';

  getClass(): string {
    const styleClasses = ['ecl-indicator', this.styleClass];

    if (this.extraClasses) {
      styleClasses.push(this.extraClasses);
    }

    return styleClasses.join(' ').trim();
  }

  componentDidRender() {
    if (this.eclScript) {
      const src = getAssetPath('./build/scripts/ecl-indicator.js');
      const existing = document.querySelector(`script[src="${src}"]`);
      if (existing) {
        existing.remove();
      }

      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const indicator = new ECL.Indicator(this.el);
        indicator.init();
      };
      document.body.appendChild(script);
    }
  }

  render() {
    return (
      <span
        class={this.getClass()} // Use the getClass method to handle the classes
        data-ecl-indicator
      >
        {this.value}
      </span>
    );
  }
}
