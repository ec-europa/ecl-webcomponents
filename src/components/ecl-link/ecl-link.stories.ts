import iconsAllEc from '@ecl/resources-ec-icons/dist/lists/all.json';

const getArgs = () => {
  return {
    variant: 'default',
    path: '/example',
    external: false,
    label: 'An ECL link',
    icon: '',
    iconPosition: 'after',
    iconTransform: '',
  };
};

const getArgTypes = () => {
  return {
    variant: {
      type: { name: 'select' },
      options: ['default', 'standalone', 'cta', 'primary', 'secondary', 'negative'],
      description: "Link variant"
    },
    label: {
      type: { name: 'string' },
      description: "Link label"
    },
    external: {
      type: { name: 'boolean' },
      description: "When the link is external"
    },
    path: {
      name: 'path/url',
      type: { name: 'string' },
      description: 'Path or Url to link to',
    },
    icon: {
      name: 'icon',
      type: { name: 'select' },
      options: iconsAllEc,
      description: "Add an icon",
    },
    iconPosition: {
      name: 'icon position',
      type: { name: 'select' },
      options: ['before', 'after'],
      description: 'Before or after the link label',
    },
    iconTransform: {
      name: 'icon transform',
      type: { name: 'select' },
      description: 'Link icon transform',
      options: [
        'rotate-90',
        'rotate-180',
        'rotate-270',
        'flip-horizontal',
        'flip-vertical',
      ],
    },
  };
};

export default {
  title: 'Components/Navigation/link',
};

const Template = (args) => 
  `<ecl-link
    variant="${args.variant}"
    theme="${args.theme}"
    path="${args.path}"
    external="${args.external}"
  >
  ${args.icon && args.iconPosition === 'before' ?
    `<ecl-icon 
      slot="icon-before"
      icon="${args.icon}"
      transform="${args.iconTransform}"
      theme="${args.theme}"
    >
    </ecl-icon>` : '' }
      ${args.label}
  ${args.icon && args.iconPosition === 'after' ?
    `<ecl-icon 
      slot="icon-after"
      icon="${args.icon}"
      theme="${args.theme}"
      transform="${args.iconTransform}"
    >
    </ecl-icon>` : '' }
  </ecl-link>`;

export const Link = Template.bind({});
Link.storyName = 'default';
Link.args = getArgs();
Link.argTypes = getArgTypes();
