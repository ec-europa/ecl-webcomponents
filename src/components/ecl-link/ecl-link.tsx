import { Component, h, Prop, Element} from '@stencil/core';

@Component({
  tag: 'ecl-link',
  styleUrls: {
    ec: './build/styles/ecl-link-ec.css',
    eu: './build/styles/ecl-link-eu.css',
  },
  shadow: false,
  scoped: false,
  assetsDirs: ['build'],
})

export class EclLink {
  @Element() el: HTMLElement;
  @Prop() theme: string = 'ec';
  @Prop() path: string;
  @Prop() styleClass: string;
  @Prop() variant: string = 'default';
  @Prop() titleAttr: string;
  @Prop() external: boolean = false;

  getClass(): string {
    const styleClasses = [
      `ecl-link`,
      `ecl-link--${this.variant}`,
      this.styleClass
    ];

    if (this.el.querySelector('ecl-icon')) {
      styleClasses.push('ecl-link--icon');
      if (this.el.querySelector('ecl-icon').getAttribute('slot') === 'icon-after') {
        styleClasses.push('ecl-link--icon-after');
      } else {
        styleClasses.push('ecl-link--icon-before');
      }
    }

    return styleClasses.join(' ');
  }

  componentDidRender() {
    if (this.el.querySelector('ecl-icon')) {
      this.el.querySelector('ecl-icon svg').classList.add('ecl-link__icon');
      if (this.el.querySelector('ecl-icon').getAttribute('slot') === 'icon-after') {
        this.el.querySelector('ecl-icon svg').classList.add('ecl-link__icon--after');
      } else {
        this.el.querySelector('ecl-icon svg').classList.add('ecl-link__icon--before');
      }
    }
  }

  getExternal() {
    return (
      <ecl-icon 
        icon="external"
        style-class={`ecl-link__icon ecl-link__icon--after sc-ecl-link-${this.theme}`}
        size = "2xs"
      >
      </ecl-icon>
    )  
  }

  render() {
    return (
      <a
        class={this.getClass()}
        href={this.path}
        title={this.titleAttr}
      >
        <slot name="icon-before"></slot>
        <slot></slot>
        <slot name="icon-after">{ this.external ? this.getExternal() : '' }</slot>
      </a>
    );
  }
}
