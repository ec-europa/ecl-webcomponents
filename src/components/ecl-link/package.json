{
  "name": "@ecl/web-components-link",
  "author": "European Commission",
  "license": "EUPL-1.2",
  "version": "0.15.1",
  "description": "ECL Link",
  "main": "ecl-link.tsx",
  "private": true,
  "scripts": {
    "build:styles": "cross-env ecl-builder styles",
    "build": "run-s clean build:*",
    "clean": "rimraf build",
    "prepublish": "npm run build"
  },
  "devDependencies": {
    "@ecl/theme-ec": "5.0.0-alpha.14",
    "@ecl/theme-eu": "5.0.0-alpha.14",
    "@ecl/builder": "5.0.0-alpha.14",
    "@ecl/resources-icons": "4.11.0",
    "@ecl/button": "5.0.0-alpha.14",
    "@ecl/icon": "5.0.0-alpha.14",
    "@ecl/link": "5.0.0-alpha.14",
    "@ecl/indicator": "5.0.0-alpha.14",
    "@ecl/utility-screen-reader": "5.0.0-alpha.14",
    "cross-env": "7.0.3"
  },
  "dependencies": {
    "@ecl/web-components-icon": "0.15.1",
    "@ecl/web-components-indicator": "0.15.1"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planctus/ecl-webcomponents.git"
  },
  "bugs": {
    "url": "https://github.com/planctus/ecl-webcomponents/issues"
  },
  "homepage": "https://github.com/planctus/ecl-webcomponents",
  "keywords": [
    "ecl",
    "europa-component-library",
    "design-system",
    "web-components"
  ]
}
