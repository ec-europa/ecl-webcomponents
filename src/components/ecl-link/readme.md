# ecl-link



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description | Type      | Default     |
| ------------ | ------------- | ----------- | --------- | ----------- |
| `external`   | `external`    |             | `boolean` | `false`     |
| `path`       | `path`        |             | `string`  | `undefined` |
| `styleClass` | `style-class` |             | `string`  | `undefined` |
| `theme`      | `theme`       |             | `string`  | `'ec'`      |
| `titleAttr`  | `title-attr`  |             | `string`  | `undefined` |
| `variant`    | `variant`     |             | `string`  | `'default'` |


## Dependencies

### Used by

 - [ecl-breadcrumb-item](../ecl-breadcrumb)
 - [ecl-category-filter-item](../ecl-category-filter)
 - [ecl-description-list-definition](../ecl-description-list)
 - [ecl-file](../ecl-file)
 - [ecl-file-translations-item](../ecl-file)
 - [ecl-hero-banner](../ecl-hero-banner)
 - [ecl-inpage-navigation-item](../ecl-inpage-navigation)
 - [ecl-news-ticker-item](../ecl-news-ticker)
 - [ecl-page-banner](../ecl-page-banner)
 - [ecl-social-media-share](../ecl-social-media-share)

### Depends on

- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-link --> ecl-icon
  ecl-breadcrumb-item --> ecl-link
  ecl-category-filter-item --> ecl-link
  ecl-description-list-definition --> ecl-link
  ecl-file --> ecl-link
  ecl-file-translations-item --> ecl-link
  ecl-hero-banner --> ecl-link
  ecl-inpage-navigation-item --> ecl-link
  ecl-news-ticker-item --> ecl-link
  ecl-page-banner --> ecl-link
  ecl-social-media-share --> ecl-link
  style ecl-link fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
