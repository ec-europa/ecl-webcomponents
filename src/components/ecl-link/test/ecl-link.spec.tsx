import { newSpecPage } from '@stencil/core/testing';
import { EclLink } from '../ecl-link';
import { EclIcon } from '../../ecl-icon/ecl-icon';
import { EclIndicator } from '../../ecl-indicator/ecl-indicator';

describe('ecl-link', () => {
  it('renders correctly as a default link', async () => {
    const page = await newSpecPage({
      components: [EclLink],
      html: `<ecl-link
              variant="default"
              path="/example.html"
            >
              A default link
            </ecl-link>`,
      });

    expect(page.root).toMatchSnapshot();
  });
  it('renders correctly as a standalone link', async () => {
    const page = await newSpecPage({
      components: [EclLink],
      html: `<ecl-link
              variant="standalone"
              path="/example.html"
            >
              A standalone link
            </ecl-link>`,
      });

    expect(page.root).toMatchSnapshot();
  });
  it('renders correctly with an icon after', async () => {
    const page = await newSpecPage({
      components: [EclLink, EclIcon],
      html: `<ecl-link
              variant="standalone"
              path="/example.html"
            >
              Link label
              <ecl-icon slot="icon-after" icon="check"> 
            </ecl-link>`,
      });

    expect(page.root).toMatchSnapshot();
  });
  it('renders correctly with an icon before', async () => {
    const page = await newSpecPage({
      components: [EclLink, EclIcon],
      html: `<ecl-link
              variant="standalone"
              theme="ec"
              path="/example.html"
            >
              Link label
              <ecl-icon slot="icon-before" icon="arrow-left"> 
            </ecl-link>`,
      });

    expect(page.root).toMatchSnapshot();
  });
  it('renders correctly when external', async () => {
    const page = await newSpecPage({
      components: [EclLink, EclIcon],
      html: `<ecl-link
              variant="standalone"
              path="/example.html"
              external
              sr-external="will move you outside of this website"
            >
              Link label
            </ecl-link>`,
      });

    expect(page.root).toMatchSnapshot();
  });
  it('renders correctly with an indicator', async () => {
    const page = await newSpecPage({
      components: [EclLink, EclIcon, EclIndicator],
      html: `<ecl-link
              variant="standalone"
              path="/example.html"
              indicator=true
              indicator-value="5"
            >
              Link label
              <ecl-icon slot="icon-after" icon="check"> 
            </ecl-link>`,
      });

    expect(page.root).toMatchSnapshot();
  });
});
