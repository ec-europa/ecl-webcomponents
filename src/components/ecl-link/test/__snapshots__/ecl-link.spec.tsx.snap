// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-link renders correctly as a default link 1`] = `
<ecl-link path="/example.html" variant="default">
  <!---->
  <a class="ecl-link ecl-link--default" href="/example.html">
    A default link
  </a>
</ecl-link>
`;

exports[`ecl-link renders correctly as a standalone link 1`] = `
<ecl-link path="/example.html" variant="standalone">
  <!---->
  <a class="ecl-link ecl-link--standalone" href="/example.html">
    A standalone link
  </a>
</ecl-link>
`;

exports[`ecl-link renders correctly when external 1`] = `
<ecl-link external="" path="/example.html" sr-external="will move you outside of this website" variant="standalone">
  <!---->
  <a class="ecl-link ecl-link--standalone" href="/example.html">
    Link label
    <slot-fb name="icon-after">
      <ecl-icon style-class="ecl-link__icon ecl-link__icon--after sc-ecl-link-ec" title-tag="will move you outside of this website">
        <svg class="ecl-icon ecl-icon--2xs ecl-link__icon ecl-link__icon--after ecl-link__icon--before sc-ecl-link-ec">
          <title>
            will move you outside of this website
          </title>
          <use xlink:href="/build/images/ec/icons.svg#external"></use>
        </svg>
      </ecl-icon>
    </slot-fb>
  </a>
</ecl-link>
`;

exports[`ecl-link renders correctly with an icon after 1`] = `
<ecl-link path="/example.html" variant="standalone">
  <!---->
  <a class="ecl-link ecl-link--icon ecl-link--icon-after ecl-link--standalone" href="/example.html">
    Link label
    <slot-fb hidden="" name="icon-after"></slot-fb>
    <ecl-icon icon="check" slot="icon-after">
      <svg class="ecl-icon ecl-icon--xs ecl-link__icon ecl-link__icon--after">
        <use xlink:href="/build/images/ec/icons.svg#check"></use>
      </svg>
    </ecl-icon>
  </a>
</ecl-link>
`;

exports[`ecl-link renders correctly with an icon before 1`] = `
<ecl-link path="/example.html" theme="ec" variant="standalone">
  <!---->
  <a class="ecl-link ecl-link--icon ecl-link--icon-before ecl-link--standalone" href="/example.html">
    <ecl-icon icon="arrow-left" slot="icon-before">
      <svg class="ecl-icon ecl-icon--xs ecl-link__icon ecl-link__icon--before">
        <use xlink:href="/build/images/ec/icons.svg#arrow-left"></use>
      </svg>
    </ecl-icon>
    Link label
  </a>
</ecl-link>
`;

exports[`ecl-link renders correctly with an indicator 1`] = `
<ecl-link indicator="true" indicator-value="5" path="/example.html" variant="standalone">
  <!---->
  <a class="ecl-link ecl-link--icon ecl-link--icon-after ecl-link--standalone" href="/example.html">
    Link label
    <span class="ecl-link__icon-container">
      <ecl-icon icon="check" slot="icon-after">
        <svg class="ecl-icon ecl-icon--xs ecl-link__icon ecl-link__icon--after">
          <use xlink:href="/build/images/ec/icons.svg#check"></use>
        </svg>
      </ecl-icon>
      <ecl-indicator ecl-script="">
        <span class="ecl-indicator" data-ecl-indicator>
          5
        </span>
      </ecl-indicator>
    </span>
  </a>
</ecl-link>
`;
